File: backend/backtesting/simulator.py
"""
Virtual portfolio simulator for backtesting.

Mirrors EBC execution thresholds from boundary/modes.py without
touching the real broker. Uses a single shared capital pool.
"""
from __future__ import annotations
from dataclasses import dataclass, field

NOTIONAL = 1000.0  # $1,000 per trade — matches live EBC config

CONFIDENCE_THRESHOLDS: dict[str, float | None] = {
    "advisory":              None,   # never execute
    "autonomous":            0.65,
    "autonomous_guardrail":  0.65,
}


@dataclass
class Position:
    ticker: str
    shares: float
    avg_cost: float
    entry_date: str


@dataclass
class VirtualPortfolio:
    initial_capital: float = 10000.0
    cash: float = field(init=False)
    positions: dict[str, Position] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.cash = self.initial_capital
        self.max_position_pct: float = 0.15

    def process(
        self,
        date: str,
        ticker: str,
        action: str,
        confidence: float,
        ebc_mode: str,
        execution_price: float | None,
        is_last_day: bool,
        confidence_threshold_override: float | None = None,
        position_value_override: float | None = None,
    ) -> dict:
        threshold = (
            confidence_threshold_override
            if confidence_threshold_override is not None
            else CONFIDENCE_THRESHOLDS.get(ebc_mode)
        )

        if threshold is None:
            return {"executed": False, "reason": "advisory_mode"}
        if is_last_day:
            return {"executed": False, "skipped_reason": "end_of_range"}
        if action == "HOLD":
            return {"executed": False, "reason": "hold_signal"}
        if confidence < threshold:
            return {"executed": False, "reason": "below_threshold"}
        if execution_price is None:
            return {"executed": False, "reason": "no_price_data"}

        if action == "BUY":
            return self._execute_buy(date, ticker, execution_price, notional=position_value_override)
        if action == "SELL":
            return self._execute_sell(ticker, execution_price)
        return {"executed": False, "reason": "unknown_action"}

    def _execute_buy(self, date: str, ticker: str, price: float, notional: float | None = None) -> dict:
        trade_notional = notional if notional is not None else NOTIONAL
        if trade_notional > 0 and price > 0:
            total_portfolio = self.cash + sum(pos.shares * pos.avg_cost for pos in self.positions.values())
            max_notional = total_portfolio * self.max_position_pct
            trade_notional = min(trade_notional, max_notional)
        if self.cash < trade_notional:
            return {"executed": False, "skipped_reason": "insufficient_funds"}
        shares = trade_notional / price
        self.cash -= trade_notional
        if ticker in self.positions:
            existing = self.positions[ticker]
            total = existing.shares + shares
            avg = (existing.shares * existing.avg_cost + shares * price) / total
            self.positions[ticker] = Position(ticker, total, avg, existing.entry_date)
        else:
            self.positions[ticker] = Position(ticker, shares, price, date)
        return {"executed": True, "action": "BUY", "shares": shares, "price": price, "notional": trade_notional}

    def _execute_sell(self, ticker: str, price: float) -> dict:
        if ticker not in self.positions:
            return {"executed": False, "reason": "no_position"}
        pos = self.positions.pop(ticker)
        proceeds = pos.shares * price
        self.cash += proceeds
        pnl = (price - pos.avg_cost) * pos.shares
        return {"executed": True, "action": "SELL", "shares": pos.shares, "price": price, "pnl": pnl}

    def portfolio_value(self, current_prices: dict[str, float]) -> float:
        position_value = sum(
            pos.shares * current_prices.get(pos.ticker, pos.avg_cost)
            for pos in self.positions.values()
        )
        return self.cash + position_value

    def mark_to_market_positions(self, current_prices: dict[str, float]) -> list[dict]:
        """Close all open positions at given prices. Mutates cash."""
        results = []
        for ticker, pos in list(self.positions.items()):
            price = current_prices.get(ticker, pos.avg_cost)
            self.cash += pos.shares * price
            pnl = (price - pos.avg_cost) * pos.shares
            results.append({"ticker": ticker, "shares": pos.shares, "price": price, "pnl": pnl, "marked_to_market": True})
        self.positions.clear()
        return results

File: lib/backtest/simulator.ts
/**
 * Virtual portfolio simulator for backtesting.
 *
 * Mirrors EBC execution thresholds from boundary/modes without touching the
 * real broker. Uses a single shared capital pool across tickers.
 *
 * Port of backend/backtesting/simulator.py.
 */

const NOTIONAL = 1000.0; // $1,000 per trade — matches live EBC config
const MAX_POSITION_PCT = 0.15; // no single ticker exceeds 15% of total portfolio at cost basis
const MIN_CASH_RESERVE_PCT = 0.1; // always keep 10% of portfolio in cash as dry powder

const CONFIDENCE_THRESHOLDS: Record<string, number | null> = {
  advisory: null, // never execute
  autonomous: 0.65,
  autonomous_guardrail: 0.65,
};

export interface Position {
  ticker: string;
  shares: number;
  avgCost: number;
  entryDate: string;
}

export interface TradeResult {
  executed: boolean;
  action?: "BUY" | "SELL";
  shares?: number;
  price?: number;
  notional?: number;
  pnl?: number;
  reason?: string;
  skippedReason?: string;
}

export class VirtualPortfolio {
  cash: number;
  positions: Map<string, Position>;
  readonly initialCapital: number;

  constructor(initialCapital = 10_000.0) {
    this.initialCapital = initialCapital;
    this.cash = initialCapital;
    this.positions = new Map();
  }

  process({
    date,
    ticker,
    action,
    confidence,
    ebcMode,
    executionPrice,
    isLastDay,
    confidenceThresholdOverride,
    positionValueOverride,
  }: {
    date: string;
    ticker: string;
    action: string;
    confidence: number;
    ebcMode: string;
    executionPrice: number | null;
    isLastDay: boolean;
    confidenceThresholdOverride?: number | null;
    positionValueOverride?: number | null;
  }): TradeResult {
    const threshold =
      confidenceThresholdOverride !== undefined && confidenceThresholdOverride !== null
        ? confidenceThresholdOverride
        : CONFIDENCE_THRESHOLDS[ebcMode] ?? null;

    if (threshold === null) return { executed: false, reason: "advisory_mode" };
    if (isLastDay) return { executed: false, skippedReason: "end_of_range" };
    if (action === "HOLD") return { executed: false, reason: "hold_signal" };
    if (confidence < threshold) return { executed: false, reason: "below_threshold" };
    if (executionPrice === null) return { executed: false, reason: "no_price_data" };

    if (action === "BUY") {
      return this._executeBuy(date, ticker, executionPrice, positionValueOverride ?? undefined);
    }
    if (action === "SELL") {
      return this._executeSell(ticker, executionPrice);
    }
    return { executed: false, reason: "unknown_action" };
  }

  private _executeBuy(
    date: string,
    ticker: string,
    price: number,
    notional?: number,
  ): TradeResult {
    let tradeNotional = notional !== undefined ? notional : NOTIONAL;

    if (tradeNotional > 0 && price > 0) {
      const totalPortfolio =
        this.cash +
        Array.from(this.positions.values()).reduce(
          (sum, pos) => sum + pos.shares * pos.avgCost,
          0,
        );

      const existing = this.positions.get(ticker);
      const existingCost = existing ? existing.shares * existing.avgCost : 0;
      const maxNotional = totalPortfolio * MAX_POSITION_PCT;
      const remainingCapacity = Math.max(0, maxNotional - existingCost);
      tradeNotional = Math.min(tradeNotional, remainingCapacity);

      const minCash = totalPortfolio * MIN_CASH_RESERVE_PCT;
      const maxSpend = Math.max(0, this.cash - minCash);
      tradeNotional = Math.min(tradeNotional, maxSpend);
    }

    if (tradeNotional <= 0) {
      return { executed: false, skippedReason: "position_cap_reached" };
    }
    if (this.cash < tradeNotional) {
      return { executed: false, skippedReason: "insufficient_funds" };
    }

    const shares = tradeNotional / price;
    this.cash -= tradeNotional;

    const existing = this.positions.get(ticker);
    if (existing) {
      const total = existing.shares + shares;
      const avgCost = (existing.shares * existing.avgCost + shares * price) / total;
      this.positions.set(ticker, { ticker, shares: total, avgCost, entryDate: existing.entryDate });
    } else {
      this.positions.set(ticker, { ticker, shares, avgCost: price, entryDate: date });
    }

    return { executed: true, action: "BUY", shares, price, notional: tradeNotional };
  }

  private _executeSell(ticker: string, price: number): TradeResult {
    const pos = this.positions.get(ticker);
    if (!pos) return { executed: false, reason: "no_position" };

    const proceeds = pos.shares * price;
    this.cash += proceeds;
    const pnl = (price - pos.avgCost) * pos.shares;
    this.positions.delete(ticker);

    return { executed: true, action: "SELL", shares: pos.shares, price, pnl };
  }

  portfolioValue(currentPrices: Record<string, number>): number {
    const positionValue = Array.from(this.positions.values()).reduce(
      (sum, pos) => sum + pos.shares * (currentPrices[pos.ticker] ?? pos.avgCost),
      0,
    );
    return this.cash + positionValue;
  }

  markToMarket(currentPrices: Record<string, number>): TradeResult[] {
    const results: TradeResult[] = [];
    for (const [ticker, pos] of this.positions) {
      const price = currentPrices[ticker] ?? pos.avgCost;
      this.cash += pos.shares * price;
      const pnl = (price - pos.avgCost) * pos.shares;
      results.push({ executed: true, action: "SELL", shares: pos.shares, price, pnl });
    }
    this.positions.clear();
    return results;
  }
}
